tools/parse/src/main/scala/org/allenai/nlpstack/parse/poly/decisiontree/RandomForest.scala

```diff
@@ -35,23 +35,32 @@ case class RandomForest(allOutcomes: Seq[Int], decisionTrees: Seq[DecisionTree])
 
   require(decisionTrees.nonEmpty, "Cannot initialize a RandomForest with zero decision trees")
 
-  /** Each decision gets a vote (i.e. suggests a distribution) about the outcome. The produced
-    * distribution is the normalized sum of the votes.
-    *
-    * @param featureVector feature vector to find outcome distribution for
-    * @return a probability distribution over outcomes
-    */
-  override def outcomeDistribution(
+  def outcomeDistribution(
     featureVector: FeatureVector
   ): (OutcomeDistribution, Option[Justification]) = {
 
-    val decisionTreeDistributions = decisionTrees map { decisionTree =>
-      decisionTree.outcomeDistribution(featureVector)._1
+    // determines whether we've seen enough of one outcome to conclude that
+    // it's the majority outcome
+    def isConfidentEnoughAboutHistogram(histogram: Map[Int, Int]): Boolean = {
+      histogram.values.max > decisionTrees.size / 2
+    }
+    var decisionTreeIndex = 0
+    var continueLoop: Boolean = true
+    var unnormalizedOutcomeDistribution = Map[Int, Int]()
+    while (decisionTreeIndex < decisionTrees.size && continueLoop) {
+      val nextDecisionTree = decisionTrees(decisionTreeIndex)
+      val outcome: (Int, Option[Justification]) = nextDecisionTree.classify(featureVector)
+      unnormalizedOutcomeDistribution =
+        unnormalizedOutcomeDistribution.updated(
+          outcome._1,
+          1 + unnormalizedOutcomeDistribution.getOrElse(outcome._1, 0)
+        )
+      continueLoop = !isConfidentEnoughAboutHistogram(unnormalizedOutcomeDistribution)
```

Comment:
Perhaps call this only if you've already traversed at least half the decision trees? Right now you have the decision tree index. If you do refactor with a `for` comprehension, you could perhaps use a `zipWithIndex`.